src/components/heartrate/Ppg.cpp

```diff
@@ -2,106 +2,276 @@
+  void Filter30to240(std::array<float, Ppg::dataLength>& signal) {
+    // From:
+    // https://www.norwegiancreations.com/2016/03/arduino-tutorial-simple-high-pass-band-pass-and-band-stop-filtering/
+
+    int length = signal.size();
+    // 0.268 is ~0.5Hz and 0.816 is ~4Hz cutoff at 10Hz sampling
+    float expAlpha = 0.816f;
+    float expAvg = 0.0f;
+    for (int loop = 0; loop < 4; loop++) {
+      expAvg = signal.front();
+      for (int idx = 0; idx < length; idx++) {
+        expAvg = (expAlpha * signal.at(idx)) + ((1 - expAlpha) * expAvg);
+        signal[idx] = expAvg;
+      }
+    }
+    expAlpha = 0.268f;
+    for (int loop = 0; loop < 4; loop++) {
+      expAvg = signal.front();
+      for (int idx = 0; idx < length; idx++) {
+        expAvg = (expAlpha * signal.at(idx)) + ((1 - expAlpha) * expAvg);
+        signal[idx] -= expAvg;
+      }
+    }
   }
 
-  int Trough(int8_t* d, size_t size, uint8_t mn, uint8_t mx) {
-    auto z2 = CompareShift(d, mn - 2, size);
-    auto z1 = CompareShift(d, mn - 1, size);
-    for (int i = mn; i < mx + 1; i++) {
-      auto z = CompareShift(d, i, size);
-      if (z2 > z1 && z1 < z) {
-        return i;
+  float SpectrumMax(const std::array<float, Ppg::spectrumLength>& data, int start, int end) {
+    float max = 0.0f;
+    for (int idx = start; idx < end; idx++) {
+      if (data.at(idx) > max) {
+        max = data.at(idx);
       }
-      z2 = z1;
-      z1 = z;
     }
-    return -1;
+    return max;
   }
-}
 
-Ppg::Ppg()
-  : hpf {0.87033078, -1.74066156, 0.87033078, -1.72377617, 0.75754694},
-    agc {20, 0.971, 2},
-    lpf {0.11595249, 0.23190498, 0.11595249, -0.72168143, 0.18549138} {
-}
+  void Detrend(std::array<float, Ppg::dataLength>& signal) {
+    int size = signal.size();
+    float offset = signal.front();
+    float slope = (signal.at(size - 1) - offset) / static_cast<float>(size - 1);
 
-int8_t Ppg::Preprocess(float spl) {
-  spl -= offset;
-  spl = hpf.Step(spl);
-  spl = agc.Step(spl);
-  spl = lpf.Step(spl);
+    for (int idx = 0; idx < size; idx++) {
+      signal[idx] -= (slope * static_cast<float>(idx) + offset);
+    }
+    for (int idx = 0; idx < size - 1; idx++) {
+      signal[idx] = signal[idx + 1] - signal[idx];
+    }
+  }
 
-  auto spl_int = static_cast<int8_t>(spl);
+  // Hanning Coefficients
+  // Note: Harcoded and must be updated if dataLength changes.
+  // This data is symetrical so just using the first half (saves 128B when dataLength is 64).
+  static constexpr float hanning[Ppg::dataLength >> 1] {
+    0.0f,        0.00248461f, 0.00991376f, 0.0222136f,  0.03926189f, 0.06088921f, 0.08688061f, 0.11697778f,
+    0.15088159f, 0.1882551f,  0.22872687f, 0.27189467f, 0.31732949f, 0.36457977f, 0.41317591f, 0.46263495f,
+    0.51246535f, 0.56217185f, 0.61126047f, 0.65924333f, 0.70564355f, 0.75f,       0.79187184f, 0.83084292f,
+    0.86652594f, 0.89856625f, 0.92664544f, 0.95048443f, 0.96984631f, 0.98453864f, 0.99441541f, 0.99937846f};
+}
```

Comment:
Where do these come from? If they're precalculated results of some calculation, can it be created using constexpr?

Reply:
@Riksu9000 I tried to use constexpr to calculate but it would not allow cos() which is needed for the calculation. Maybe it will be allowed in the future? If so, I will update to more elegant and versatile constexpr technique.

Reply:
Some mathematical functions are made `constexpr` with C++23, but that does not include `cos()` or `sin()`.

Reply:
Left it "as is" for now. Note that changing dataLength will break it in the compiler, so should be low risk of future developer error.